nautilus_core/indicators/src/momentum/bb.rs

```diff
@@ -0,0 +1,223 @@
+    pub period: usize,
+    pub k: f64,
+    pub ma_type: MovingAverageType,
+    pub upper: f64,
+    pub middle: f64,
+    pub lower: f64,
+    pub initialized: bool,
+    ma: Box<dyn MovingAverage + Send + 'static>,
+    prices: VecDeque<f64>,
+    has_inputs: bool,
+}
+
+impl Display for BollingerBands {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}({},{},{})",
+            self.name(),
+            self.period,
+            self.k,
+            self.ma_type,
+        )
+    }
+}
+
+impl Indicator for BollingerBands {
+    fn name(&self) -> String {
+        stringify!(BollingerBands).to_string()
+    }
+
+    fn has_inputs(&self) -> bool {
+        self.has_inputs
+    }
+
+    fn initialized(&self) -> bool {
+        self.initialized
+    }
+
+    fn handle_quote_tick(&mut self, tick: &QuoteTick) {
+        let bid = tick.bid_price.raw as f64;
+        let ask = tick.ask_price.raw as f64;
+        let mid = (bid + ask) / 2.0;
+        self.update_raw(ask, bid, mid);
+    }
+
+    fn handle_trade_tick(&mut self, tick: &TradeTick) {
+        let price = tick.price.raw as f64;
+        self.update_raw(price, price, price);
+    }
+
+    fn handle_bar(&mut self, bar: &Bar) {
+        self.update_raw((&bar.high).into(), (&bar.low).into(), (&bar.close).into());
+    }
+
+    fn reset(&mut self) {
+        self.ma.reset();
+        self.prices.clear();
+        self.upper = 0.0;
+        self.middle = 0.0;
+        self.lower = 0.0;
+        self.has_inputs = false;
+        self.initialized = false;
+    }
+}
+
+impl BollingerBands {
+    /// Creates a new [`BollingerBands`] instance.
+    pub fn new(period: usize, k: f64, ma_type: Option<MovingAverageType>) -> anyhow::Result<Self> {
+        Ok(Self {
+            period,
+            k,
+            ma_type: ma_type.unwrap_or(MovingAverageType::Simple),
+            has_inputs: false,
+            initialized: false,
+            upper: 0.0,
+            middle: 0.0,
+            lower: 0.0,
+            ma: MovingAverageFactory::create(ma_type.unwrap_or(MovingAverageType::Simple), period),
+            prices: VecDeque::with_capacity(period),
+        })
+    }
+
+    pub fn update_raw(&mut self, high: f64, low: f64, close: f64) {
+        let typical = (high + low + close) / 3.0;
+        self.prices.push_back(typical);
+        self.ma.update_raw(typical);
+
+        // Initialization logic
+        if !self.initialized {
+            self.has_inputs = true;
+            if self.prices.len() >= self.period {
+                self.initialized = true;
+            }
+        }
+
+        // Calculate values
+        let std = fast_std_with_mean(self.prices.clone(), self.ma.value());
+
+        self.upper = self.ma.value() + self.k * std;
+        self.middle = self.ma.value();
+        self.lower = self.ma.value() - self.k * std;
+    }
+}
+
+pub fn fast_std_with_mean(values: VecDeque<f64>, mean: f64) -> f64 {
```

Comment:
What would be the best place to keep this? Because in the future, few indicators need this.